### PR TITLE
[CompilerRT] Add numerical sanitizer

### PR DESCRIPTION
**Pull request by alexander-shaposhnikov**
This PR contains the compiler-rt changes that were split out from https://github.com/llvm/llvm-project/pull/85916.
Follow-up patch with numerical tests will be sent once the initial changes (instrumentation, clang, runtime) land.

Test plan:

1. cd build/runtimes/runtimes-bins && ninja check-nsan
2. ninja check-all

### CODE REVIEW

**Comment by vitalybuka** [comment]
I assume, this one depends on #85916 ?

File: compiler-rt/lib/nsan/nsan_suppressions.cc [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        Suppressions->Match(Cur->info.file, SupprType, &S)) {
+      break;
+    }
+  }
+  Frames->ClearAll();
+  return S;
+}
+
+Suppression *GetSuppressionForStack(const StackTrace *Stack,
+                                    const char *SupprType) {
```
Comment (vitalybuka) [inline compiler-rt/lib/nsan/nsan_suppressions.cc:66]: can this be an enum, to hide kSuppression* from the header?

File: compiler-rt/test/nsan/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    ${CMAKE_CURRENT_BINARY_DIR}/Unit/lit.site.cfg.py)
+  list(APPEND NSAN_TEST_DEPS NsanUnitTests)
+  list(APPEND NSAN_TESTSUITES ${CMAKE_CURRENT_BINARY_DIR}/Unit)
+endif()
+
+add_lit_testsuite(check-nsan "Running the numerical stability sanitizer tests"
+  ${NSAN_TESTSUITES}
+  DEPENDS ${NSAN_TEST_DEPS}
+  )
+set_target_properties(check-nsan PROPERTIES FOLDER "Compiler-RT Misc")
```
Comment (vitalybuka) [inline compiler-rt/test/nsan/CMakeLists.txt:20]: https://github.com/llvm/llvm-project/pull/89753/files

File: compiler-rt/lib/nsan/tests/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  -g
+  -fno-omit-frame-pointer)
+
+file(GLOB NSAN_HEADERS ../*.h)
+set(NSAN_UNITTESTS
+	NSanUnitTest.cpp
+        nsan_unit_test_main.cpp)
+
+add_custom_target(NsanUnitTests)
+set_target_properties(NsanUnitTests PROPERTIES FOLDER "Compiler-RT Tests")
```
Comment (vitalybuka) [inline compiler-rt/lib/nsan/tests/CMakeLists.txt:18]: https://github.com/llvm/llvm-project/pull/89753/files

File: compiler-rt/lib/nsan/nsan_suppressions.cc (hunk 1 above)
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/nsan/nsan_suppressions.cc]: Done

File: compiler-rt/test/nsan/CMakeLists.txt (hunk 2 above)
Comment (alexander-shaposhnikov) [reply compiler-rt/test/nsan/CMakeLists.txt]: Done

File: compiler-rt/lib/nsan/tests/CMakeLists.txt (hunk 3 above)
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/nsan/tests/CMakeLists.txt]: Done

File: compiler-rt/lib/nsan/nsan.h [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+namespace __nsan {
+
+extern bool NsanInitialized;
+extern bool NsanInitIsRunning;
+
+void initializeInterceptors();
+
+// See notes in nsan_platform.
+// printf-free (see comment in nsan_interceptors.cc).
+inline char *getShadowAddrFor(char *Ptr) {
```
Comment (vitalybuka) [inline compiler-rt/lib/nsan/nsan.h:60]: u8* and the rest?

File: compiler-rt/lib/nsan/nsan_stats.cc [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+template <typename MapT, typename VectorT, typename Fn>
+void UpdateEntry(CheckTypeT CheckTy, uptr PC, uptr BP, MapT *Map,
+                 VectorT *Vector, Mutex *Mutex, Fn F) {
+  BufferedStackTrace Stack;
+  Stack.Unwind(PC, BP, nullptr, false);
+  u32 StackId = StackDepotPut(Stack);
+  typename MapT::Handle Handle(Map, key(CheckTy, StackId));
+  Lock L(Mutex);
```
Comment (vitalybuka) [inline compiler-rt/lib/nsan/nsan_stats.cc:48]: AddrHashMap is for limited lock access 

If it always locks, we can just use DenseMap<>, which has simpler interface.

File: compiler-rt/lib/nsan/nsan_platform.h [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+//       unknown.
+//       This allows handling unaligned loat load/stores by checking that a load
+//       with a given alignment corresponds to the alignment of the store.
+//       Any store of a non-floating point type invalidates the corresponding
+//       bytes, so that subsequent overlapping loads (aligned or not) know that
+//       the corresponding shadow value is no longer valid.
+
+// On Linux/x86_64, memory is laid out as follows:
+//
+// +--------------------+ 0x800000000000 (top of memory)
```
Comment (vitalybuka) [inline compiler-rt/lib/nsan/nsan_platform.h:37]: Maybe @thurstond can provide a feedback for these mappings?

**Review by vitalybuka — APPROVED** [review]

File: compiler-rt/lib/nsan/nsan.h (hunk 4 above)
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/nsan/nsan.h]: Done

File: compiler-rt/lib/nsan/nsan_stats.cc (hunk 5 above)
Comment (alexander-shaposhnikov) [reply compiler-rt/lib/nsan/nsan_stats.cc:48]: unlike DenseMap AddrHashMap uses MmapOrDie* for memory allocation and avoids any deps on libc.

**Comment by MaskRay** [comment]
I am sorry that I did not notice this patch but I feel sad that this has been committed with many style issues. I fully support numerical sanitizer but I think we really should do a better code style job for posterity.  Created #96142

**Comment by alexander-shaposhnikov** [comment]
Ack

**Comment by ilovepi** [comment]
Hi, we're seeing some test failures in our Linux CI bots for arm64 and x64 after this patch. It appears that the c++ file for gtest is being compiled w/ `clang` instead of `clang++`. Its not clear to me why that's happening, from a quick look at the patch. Probably a small mistake in the CMake?

Error Message:
```
FAILED: compiler-rt/lib/nsan/tests/NsanTestObjects.gtest-all.cc.x86_64.o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/lib/nsan/tests/NsanTestObjects.gtest-all.cc.x86_64.o 
cd /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/lib/nsan/tests && /b/s/w/ir/x/w/llvm_build/./bin/clang --target=x86_64-unknown-linux-gnu -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -Wctad-maybe-unsupported -fdiagnostics-color -ffunction-sections -fdata-sections -ffile-prefix-map=/b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins=../../../llvm-llvm-project -ffile-prefix-map=/b/s/w/ir/x/w/llvm-llvm-project/= -no-canonical-prefixes -Wall -Wno-unused-parameter -Wno-unknown-warning-option -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta -I/b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/include -nostdinc++ -g -Wno-covered-switch-default -Wno-suggest-override -DGTEST_NO_LLVM_SUPPORT=1 -DGTEST_HAS_RTTI=0 -I/b/s/w/ir/x/w/llvm-llvm-project/runtimes/../third-party/unittest/googletest/include -I/b/s/w/ir/x/w/llvm-llvm-project/runtimes/../third-party/unittest/googletest -I/b/s/w/ir/x/w/llvm-llvm-project/compiler-rt/lib/ -DSANITIZER_COMMON_REDEFINE_BUILTINS_IN_STD -O2 -g -fno-omit-frame-pointer -c -o NsanTestObjects.gtest-all.cc.x86_64.o /b/s/w/ir/x/w/llvm-llvm-project/third-party/unittest/googletest/src/gtest-all.cc
In file included from /b/s/w/ir/x/w/llvm-llvm-project/third-party/unittest/googletest/src/gtest-all.cc:38:
/b/s/w/ir/x/w/llvm-llvm-project/runtimes/../third-party/unittest/googletest/include/gtest/gtest.h:52:10: fatal error: 'cstddef' file not found
   52 | #include <cstddef>
      |          ^~~~~~~~~
1 error generated.
```

Failing bot:   https://ci.chromium.org/ui/p/fuchsia/builders/toolchain.ci/clang-linux-x64/b8744662427054892657/overview

Bug: https://fxbug.dev/348377270

I'm not sure how this is passing in buildbot. Can you take a look, and if its hard to fix quickly, revert until it can be relanded?

As an FYI on our build config, it isn't special, its a normal Linux toolchain, using the RUNTIMES build, and fails w/ both stage1 and stage2 toolchains. Of note is that our bots don't have a libc installed, and we build our toolchains hermetically w/ a curated Linux sysroot. There's a chance your patch isn't respecting the SYSROOT correctly, but I haven't looked closely enough to be sure.

That said, I'm fairly certain the error in this case is using the c compiler vs. the c++ compiler.

**Comment by alexander-shaposhnikov** [comment]
thanks for the report, I'll look shortly (if don't find - will revert for now)

**Comment by ilovepi** [comment]
TY!

File: compiler-rt/lib/nsan/tests/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+  -fno-omit-frame-pointer)
+
+file(GLOB NSAN_HEADERS ../*.h)
+set(NSAN_UNITTESTS
+	NSanUnitTest.cpp
+        nsan_unit_test_main.cpp)
+
+add_custom_target(NsanUnitTests)
+
+# set(NSAN_UNITTEST_LINK_FLAGS ${COMPILER_RT_UNITTEST_LINK_FLAGS} -ldl)
```
Comment (ornata) [inline compiler-rt/lib/nsan/tests/CMakeLists.txt:19]: Should these lines be commented out?

**Pull request merged**